File: src/binds/c/framebuffer/framebuffer.c
#include "vulkan/vulkan_core.h"
#include <stdint.h>
#include <vulkan/vulkan.h>
#include <stdlib.h>


typedef struct framebufferInitResult {
	VkResult result;
	VkFramebuffer* framebuffers;
	uint16_t len;
} framebufferInitResult;

framebufferInitResult framebufferInit(VkImageView* views, int num_views, VkDevice dev, VkRenderPass renderpass, VkExtent2D extent) {
	framebufferInitResult result;
	result.framebuffers = malloc(num_views * sizeof(VkFramebuffer));
	result.len = num_views;
	for (int i = 0 ; i < num_views; i++) {
		VkImageView* attachment = &views[i];
		VkFramebufferCreateInfo create_info = {
			.sType = VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO,
			.renderPass = renderpass,
			.attachmentCount = 1,
			.pAttachments = attachment,
			.width = extent.width,
			.height = extent.height,
			.layers = 1,
		};
		VkResult res = vkCreateFramebuffer(dev, &create_info, NULL, &result.framebuffers[i]);
		if (res != VK_SUCCESS) {
			result.result = res;
			return result;
		}
	}
	return result;
}


void framebufferDeinit(VkDevice dev, framebufferInitResult result) {
	for (int i = 0; i < result.len; i++) {
		vkDestroyFramebuffer(dev, result.framebuffers[i], NULL);
	}
}

File: src/binds/c/framebuffer/framebuffer.zig
const vk = @import("../vulkan.zig");


const framebufferInitResult = extern struct {
    result: vk.VkResult,
    framebuffers: [*c] vk.VkFramebuffer,
    len: u32,
};

extern fn framebufferInit([*c]vk.VkImageView, c_int, vk.VkDevice, vk.VkRenderPass, vk.VkExtent2D) framebufferInitResult;
extern fn framebufferDeinit(vk.VkDevice, framebufferInitResult) void;


pub const errors = error{
    framebufferInitializationFailed
};


pub const framebuffer = struct {
    framebuffers: []vk.VkFramebuffer,
    result: framebufferInitResult,
    device: vk.VkDevice,

    pub fn init(views: []vk.VkImageView, dev: vk.VkDevice, renderpass: vk.VkRenderPass, extent: vk.VkExtent2D) !framebuffer {
        var fb: framebuffer = undefined;
        try fb._init(views, dev, renderpass, extent);
        return fb;
    }

    fn _init(self: *framebuffer, views: []vk.VkImageView, dev: vk.VkDevice, renderpass: vk.VkRenderPass, extent: vk.VkExtent2D) !void {
        const result = framebufferInit(views.ptr, @intCast(views.len), dev, renderpass, extent);
        if (result.result != vk.VK_SUCCESS) return errors.framebufferInitializationFailed;
        self.framebuffers.ptr = result.framebuffers;
        self.framebuffers.len = result.len;
        self.result = result;
        self.device = dev;
    }

    pub fn deinit(self: *framebuffer) void {
        framebufferDeinit(self.device, self.result);
    }
};
